File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/type/NativeTypeProviderRegistry.swift
import Foundation

internal class NativeTypeProviderRegistry {
    private static var providerMap: [String: NativeTypeProvider] = [:]
    private static let lock = NSLock()

    static func getOrCreate(_ instanceName: String) -> NativeTypeProvider {
        lock.lock()
        defer { lock.unlock() }

        if let existingProvider = providerMap[instanceName] {
            return existingProvider
        } else {
            let newProvider = NativeTypeProvider()
            providerMap[instanceName] = newProvider
            return newProvider
        }
    }

    static func remove(_ instanceName: String) {
        lock.lock()
        defer { lock.unlock() }
        providerMap.removeValue(forKey: instanceName)
    }
}

File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/type/NativeTypeProviderRegistry.kt
package io.nativeblocks.runtime.api.provider.type

internal object NativeTypeProviderRegistry {

    private val providerMap = mutableMapOf<String, NativeTypeProvider>()
    private val lock = Any()

    fun getOrCreate(instanceName: String): NativeTypeProvider {
        synchronized(lock) {
            return providerMap.getOrPut(instanceName) {
                NativeTypeProvider()
            }
        }
    }

    fun remove(instanceName: String) {
        synchronized(lock) {
            providerMap.remove(instanceName)
        }
    }

}